opencda/core/ml_libs/reinforcement_learning/envs/simple_carla_env.py

```diff
@@ -0,0 +1,726 @@
+    Carla server and gets running status. The observation is obtained from simulator's state, information and
+    sensor data, along with reward which can be calculated and retrived.
+
+    When created, it will initialize environment with config and Carla TCP host & port. This method will NOT create
+    simulator instance. It only creates some data structures to store information when running env.
+
+    :Arguments:
+        - cfg (Dict): Env config dict.
+        - host (str, optional): Carla server IP host. Defaults to 'localhost'.
+        - port (int, optional): Carla server IP port. Defaults to 9000.
+        - tm_port (Optional[int], optional): Carla Traffic Manager port. Defaults to None.
+
+    :Interfaces: reset, step, close, is_success, is_failure, render, seed
+
+    :Properties:
+        - hero_player (carla.Actor): Hero vehicle in simulator.
+    """
+
+    metadata = {'render.modes': ['rgb_array']}
+    action_space = spaces.Dict({})
+    observation_space = spaces.Dict({})
+    reward_type = ['goal', 'distance', 'speed', 'angle', 'steer', 'lane', 'failure']
+    config = dict(
+        simulator=dict(),
+        # reward types total reward take into account
+        reward_type=['goal', 'distance', 'speed', 'angle', 'failure'],
+        # reward value if success
+        success_reward=10,
+        # failure judgement
+        col_is_failure=False,
+        stuck_is_failure=False,
+        ignore_light=False,
+        ran_light_is_failure=False,
+        off_road_is_failure=False,
+        wrong_direction_is_failure=False,
+        off_route_is_failure=False,
+        # failure judgement hyper-parameters
+        off_route_distance=6,
+        success_distance=5,
+        stuck_len=300,
+        max_speed=5,
+        # whether open visualize
+        visualize=None,
+    )
+
+    def __init__(
+            self,
+            cfg: Dict,
+            host: str = 'localhost',
+            port: int = 9000,
+            tm_port: Optional[int] = None,
+            carla_timeout: Optional[int] = 60.0,
+            **kwargs,
+    ) -> None:
+
+        # ------------ copy from base class -------------
+        if 'cfg_type' not in cfg:
+            self._cfg = self.__class__.default_config()
+            self._cfg = deep_merge_dicts(self._cfg, cfg)
+        else:
+            self._cfg = cfg
+        utils.EzPickle.__init__(self)
+        # ----------------------------------------------
+        """
+        Initialize environment with config and Carla TCP host & port.
+        """
+        super().__init__(cfg, **kwargs)
+
+        self._simulator_cfg = self._cfg.simulator
+        self._carla_host = host
+        self._carla_port = port
+        self._carla_tm_port = tm_port
+        self._carla_timeout = carla_timeout
+
+        self._use_local_carla = False
+        if self._carla_host != 'localhost':
+            self._use_local_carla = True
+        self._simulator = None
+
+        self._col_is_failure = self._cfg.col_is_failure
```

Comment:
please give comments on what's are these variable used for, otherwise we can not understand